// langium/lib/grammar/validation/validation-resources-collector.js
"use strict";
/******************************************************************************
 * Copyright 2022 TypeFox GmbH
 * This program and the accompanying materials are made available under the
 * terms of the MIT License, which is available in the project root.
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.LangiumGrammarValidationResourcesCollector = void 0;
const collections_1 = require("../../utils/collections");
const stream_1 = require("../../utils/stream");
const ast_1 = require("../generated/ast");
const internal_grammar_util_1 = require("../internal-grammar-util");
const types_util_1 = require("../type-system/types-util");
const ast_collector_1 = require("../type-system/ast-collector");
class LangiumGrammarValidationResourcesCollector {
    constructor(services) {
        this.documents = services.shared.workspace.LangiumDocuments;
    }
    collectValidationResources(grammar) {
        const typeResources = (0, ast_collector_1.collectValidationAst)(grammar, this.documents);
        return {
            typeToValidationInfo: this.collectValidationInfo(typeResources),
            typeToSuperProperties: this.collectSuperProperties(typeResources),
        };
    }
    collectValidationInfo({ astResources, inferred, declared }) {
        const res = new Map();
        const typeNameToRulesActions = collectNameToRulesActions(astResources);
        for (const type of (0, types_util_1.mergeTypesAndInterfaces)(inferred)) {
            res.set(type.name, { inferred: type, inferredNodes: typeNameToRulesActions.get(type.name) });
        }
        const typeNametoInterfacesUnions = (0, stream_1.stream)(astResources.interfaces)
            .concat(astResources.types)
            .reduce((acc, type) => acc.set(type.name, type), new Map());
        for (const type of (0, types_util_1.mergeTypesAndInterfaces)(declared)) {
            const node = typeNametoInterfacesUnions.get(type.name);
            if (node) {
                const inferred = res.get(type.name);
                res.set(type.name, Object.assign(Object.assign({}, inferred !== null && inferred !== void 0 ? inferred : {}), { declared: type, declaredNode: node }));
            }
        }
        return res;
    }
    collectSuperProperties({ inferred, declared }) {
        const typeToSuperProperties = new Map();
        const interfaces = (0, types_util_1.mergeInterfaces)(inferred, declared);
        const interfaceMap = new Map(interfaces.map(e => [e.name, e]));
        for (const type of (0, types_util_1.mergeInterfaces)(inferred, declared)) {
            typeToSuperProperties.set(type.name, this.addSuperProperties(type, interfaceMap, new Set()));
        }
        return typeToSuperProperties;
    }
    addSuperProperties(interfaceType, map, visited) {
        if (visited.has(interfaceType.name)) {
            return [];
        }
        visited.add(interfaceType.name);
        const properties = [...interfaceType.properties];
        for (const superType of interfaceType.superTypes) {
            const value = map.get(superType.name);
            if (value) {
                properties.push(...this.addSuperProperties(value, map, visited));
            }
        }
        return properties;
    }
}
exports.LangiumGrammarValidationResourcesCollector = LangiumGrammarValidationResourcesCollector;
function collectNameToRulesActions({ parserRules, datatypeRules }) {
    const acc = new collections_1.MultiMap();
    // collect rules
    (0, stream_1.stream)(parserRules)
        .concat(datatypeRules)
        .forEach(rule => acc.add((0, internal_grammar_util_1.getRuleType)(rule), rule));
    // collect actions
    function collectActions(element) {
        if ((0, ast_1.isAction)(element)) {
            const name = (0, internal_grammar_util_1.getActionType)(element);
            if (name) {
                acc.add(name, element);
            }
        }
        if ((0, ast_1.isAlternatives)(element) || (0, ast_1.isGroup)(element) || (0, ast_1.isUnorderedGroup)(element)) {
            element.elements.forEach(e => collectActions(e));
        }
    }
    parserRules
        .forEach(rule => collectActions(rule.definition));
    return acc;
}
