// zenstack/utils/ast-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllLoadedAndReachableDataModels = exports.getAllLoadedDataModels = exports.findUpAst = exports.getContainingDataModel = exports.isCollectionPredicate = exports.getAllDataModelsIncludingImports = exports.getAllDeclarationsIncludingImports = exports.resolveImport = exports.resolveTransitiveImports = exports.resolveImportUri = exports.getDataModelFieldReference = exports.isFutureInvocation = exports.isAuthInvocation = exports.getIdFields = exports.mergeBaseModels = exports.extractDataModelsWithAllowRules = void 0;
const ast_1 = require("@zenstackhq/language/ast");
const sdk_1 = require("@zenstackhq/sdk");
const langium_1 = require("langium");
const node_path_1 = require("node:path");
const vscode_uri_1 = require("vscode-uri");
const pkg_utils_1 = require("./pkg-utils");
function extractDataModelsWithAllowRules(model) {
    return model.declarations.filter((d) => (0, ast_1.isDataModel)(d) && d.attributes.some((attr) => { var _a; return ((_a = attr.decl.ref) === null || _a === void 0 ? void 0 : _a.name) === '@@allow'; }));
}
exports.extractDataModelsWithAllowRules = extractDataModelsWithAllowRules;
function mergeBaseModels(model, linker) {
    const buildReference = linker.buildReference.bind(linker);
    model.declarations.filter(ast_1.isDataModel).forEach((dataModel) => {
        const bases = (0, sdk_1.getRecursiveBases)(dataModel).reverse();
        if (bases.length > 0) {
            dataModel.fields = bases
                .flatMap((base) => base.fields)
                // don't inherit skip-level fields
                .filter((f) => !f.$inheritedFrom)
                .map((f) => cloneAst(f, dataModel, buildReference))
                .concat(dataModel.fields);
            dataModel.attributes = bases
                .flatMap((base) => base.attributes.filter((attr) => filterBaseAttribute(base, attr)))
                .map((attr) => cloneAst(attr, dataModel, buildReference))
                .concat(dataModel.attributes);
        }
        // mark base merged
        dataModel.$baseMerged = true;
    });
    // remove abstract models
    model.declarations = model.declarations.filter((x) => !((0, ast_1.isDataModel)(x) && x.isAbstract));
    model.declarations.filter(ast_1.isDataModel).forEach((dm) => {
        // remove abstract super types
        dm.superTypes = dm.superTypes.filter((t) => t.ref && (0, sdk_1.isDelegateModel)(t.ref));
        // fix $containerIndex
        (0, langium_1.linkContentToContainer)(dm);
    });
    // fix $containerIndex after deleting abstract models
    (0, langium_1.linkContentToContainer)(model);
}
exports.mergeBaseModels = mergeBaseModels;
function filterBaseAttribute(base, attr) {
    if (attr.$inheritedFrom) {
        // don't inherit from skip-level base
        return false;
    }
    // uninheritable attributes for all inheritance
    const uninheritableAttributes = ['@@delegate', '@@map'];
    // uninheritable attributes for delegate inheritance (they reference fields from the base)
    const uninheritableFromDelegateAttributes = ['@@unique', '@@index', '@@fulltext'];
    if (uninheritableAttributes.includes(attr.decl.$refText)) {
        return false;
    }
    if ((0, sdk_1.isDelegateModel)(base) && uninheritableFromDelegateAttributes.includes(attr.decl.$refText)) {
        return false;
    }
    return true;
}
// deep clone an AST, relink references, and set its container
function cloneAst(node, newContainer, buildReference) {
    var _a;
    const clone = (0, langium_1.copyAstNode)(node, buildReference);
    clone.$container = newContainer;
    clone.$inheritedFrom = (_a = node.$inheritedFrom) !== null && _a !== void 0 ? _a : (0, langium_1.getContainerOfType)(node, ast_1.isDataModel);
    return clone;
}
function getIdFields(dataModel) {
    var _a;
    const fieldLevelId = (0, sdk_1.getModelFieldsWithBases)(dataModel).find((f) => f.attributes.some((attr) => attr.decl.$refText === '@id'));
    if (fieldLevelId) {
        return [fieldLevelId];
    }
    else {
        // get model level @@id attribute
        const modelIdAttr = dataModel.attributes.find((attr) => { var _a, _b; return ((_b = (_a = attr.decl) === null || _a === void 0 ? void 0 : _a.ref) === null || _b === void 0 ? void 0 : _b.name) === '@@id'; });
        if (modelIdAttr) {
            // get fields referenced in the attribute: @@id([field1, field2]])
            if (!(0, ast_1.isArrayExpr)((_a = modelIdAttr.args[0]) === null || _a === void 0 ? void 0 : _a.value)) {
                return [];
            }
            const argValue = modelIdAttr.args[0].value;
            return argValue.items
                .filter((expr) => (0, ast_1.isReferenceExpr)(expr) && !!getDataModelFieldReference(expr))
                .map((expr) => expr.target.ref);
        }
    }
    return [];
}
exports.getIdFields = getIdFields;
function isAuthInvocation(node) {
    var _a;
    return (0, ast_1.isInvocationExpr)(node) && ((_a = node.function.ref) === null || _a === void 0 ? void 0 : _a.name) === 'auth' && (0, sdk_1.isFromStdlib)(node.function.ref);
}
exports.isAuthInvocation = isAuthInvocation;
function isFutureInvocation(node) {
    var _a;
    return (0, ast_1.isInvocationExpr)(node) && ((_a = node.function.ref) === null || _a === void 0 ? void 0 : _a.name) === 'future' && (0, sdk_1.isFromStdlib)(node.function.ref);
}
exports.isFutureInvocation = isFutureInvocation;
function getDataModelFieldReference(expr) {
    if ((0, ast_1.isReferenceExpr)(expr) && (0, ast_1.isDataModelField)(expr.target.ref)) {
        return expr.target.ref;
    }
    else if ((0, ast_1.isMemberAccessExpr)(expr) && (0, ast_1.isDataModelField)(expr.member.ref)) {
        return expr.member.ref;
    }
    else {
        return undefined;
    }
}
exports.getDataModelFieldReference = getDataModelFieldReference;
function resolveImportUri(imp) {
    var _a;
    if (!imp.path)
        return undefined; // This will return true if imp.path is undefined, null, or an empty string ("").
    if (!imp.path.endsWith('.zmodel')) {
        imp.path += '.zmodel';
    }
    if (!imp.path.startsWith('.') && // Respect relative paths
        !(0, node_path_1.isAbsolute)(imp.path) // Respect Absolute paths
    ) {
        imp.path = (_a = (0, pkg_utils_1.findNodeModulesFile)(imp.path)) !== null && _a !== void 0 ? _a : imp.path;
    }
    const dirUri = vscode_uri_1.Utils.dirname((0, langium_1.getDocument)(imp).uri);
    return vscode_uri_1.Utils.resolvePath(dirUri, imp.path);
}
exports.resolveImportUri = resolveImportUri;
function resolveTransitiveImports(documents, model) {
    return resolveTransitiveImportsInternal(documents, model);
}
exports.resolveTransitiveImports = resolveTransitiveImports;
function resolveTransitiveImportsInternal(documents, model, initialModel = model, visited = new Set(), models = new Set()) {
    const doc = (0, langium_1.getDocument)(model);
    const initialDoc = (0, langium_1.getDocument)(initialModel);
    if (initialDoc.uri.fsPath.toLowerCase() !== doc.uri.fsPath.toLowerCase()) {
        models.add(model);
    }
    const normalizedPath = doc.uri.fsPath.toLowerCase();
    if (!visited.has(normalizedPath)) {
        visited.add(normalizedPath);
        for (const imp of model.imports) {
            const importedModel = resolveImport(documents, imp);
            if (importedModel) {
                resolveTransitiveImportsInternal(documents, importedModel, initialModel, visited, models);
            }
        }
    }
    return Array.from(models);
}
function resolveImport(documents, imp) {
    const resolvedUri = resolveImportUri(imp);
    try {
        if (resolvedUri) {
            const resolvedDocument = documents.getOrCreateDocument(resolvedUri);
            const node = resolvedDocument.parseResult.value;
            if ((0, ast_1.isModel)(node)) {
                return node;
            }
        }
    }
    catch (_a) {
        // NOOP
    }
    return undefined;
}
exports.resolveImport = resolveImport;
function getAllDeclarationsIncludingImports(documents, model) {
    const imports = resolveTransitiveImports(documents, model);
    return model.declarations.concat(...imports.map((imp) => imp.declarations));
}
exports.getAllDeclarationsIncludingImports = getAllDeclarationsIncludingImports;
function getAllDataModelsIncludingImports(documents, model) {
    return getAllDeclarationsIncludingImports(documents, model).filter(ast_1.isDataModel);
}
exports.getAllDataModelsIncludingImports = getAllDataModelsIncludingImports;
function isCollectionPredicate(node) {
    return (0, ast_1.isBinaryExpr)(node) && ['?', '!', '^'].includes(node.operator);
}
exports.isCollectionPredicate = isCollectionPredicate;
function getContainingDataModel(node) {
    let curr = node.$container;
    while (curr) {
        if ((0, ast_1.isDataModel)(curr)) {
            return curr;
        }
        curr = curr.$container;
    }
    return undefined;
}
exports.getContainingDataModel = getContainingDataModel;
/**
 * Walk upward from the current AST node to find the first node that satisfies the predicate.
 */
function findUpAst(node, predicate) {
    let curr = node;
    while (curr) {
        if (predicate(curr)) {
            return curr;
        }
        curr = curr.$container;
    }
    return undefined;
}
exports.findUpAst = findUpAst;
/**
 * Gets all data models from all loaded documents
 */
function getAllLoadedDataModels(langiumDocuments) {
    return langiumDocuments.all
        .map((doc) => doc.parseResult.value)
        .flatMap((model) => model.declarations.filter(ast_1.isDataModel))
        .toArray();
}
exports.getAllLoadedDataModels = getAllLoadedDataModels;
/**
 * Gets all data models from loaded and reachable documents
 */
function getAllLoadedAndReachableDataModels(langiumDocuments, fromModel) {
    // get all data models from loaded documents
    const allDataModels = getAllLoadedDataModels(langiumDocuments);
    if (fromModel) {
        // merge data models transitively reached from the current model
        const model = (0, langium_1.getContainerOfType)(fromModel, ast_1.isModel);
        if (model) {
            const transitiveDataModels = getAllDataModelsIncludingImports(langiumDocuments, model);
            transitiveDataModels.forEach((dm) => {
                if (!allDataModels.includes(dm)) {
                    allDataModels.push(dm);
                }
            });
        }
    }
    return allDataModels;
}
exports.getAllLoadedAndReachableDataModels = getAllLoadedAndReachableDataModels;
