// @zenstackhq/server/api/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerCustomSerializers = exports.logInfo = exports.logWarning = exports.logError = void 0;
const decimal_js_1 = require("decimal.js");
const superjson_1 = __importDefault(require("superjson"));
function logError(logger, message, code) {
    if (logger === undefined) {
        console.error(`@zenstackhq/server: error ${code ? '[' + code + ']' : ''}, ${message}`);
    }
    else if (logger === null || logger === void 0 ? void 0 : logger.error) {
        logger.error(message, code);
    }
}
exports.logError = logError;
function logWarning(logger, message) {
    if (logger === undefined) {
        console.warn(`@zenstackhq/server: ${message}`);
    }
    else if (logger === null || logger === void 0 ? void 0 : logger.warn) {
        logger.warn(message);
    }
}
exports.logWarning = logWarning;
function logInfo(logger, message) {
    if (logger === undefined) {
        console.log(`@zenstackhq/server: ${message}`);
    }
    else if (logger === null || logger === void 0 ? void 0 : logger.info) {
        logger.info(message);
    }
}
exports.logInfo = logInfo;
/**
 * Registers custom superjson serializers.
 */
function registerCustomSerializers() {
    superjson_1.default.registerCustom({
        isApplicable: (v) => decimal_js_1.Decimal.isDecimal(v),
        serialize: (v) => v.toJSON(),
        deserialize: (v) => new decimal_js_1.Decimal(v),
    }, 'Decimal');
    // `Buffer` is not available in edge runtime
    if (globalThis.Buffer) {
        superjson_1.default.registerCustom({
            isApplicable: (v) => Buffer.isBuffer(v),
            serialize: (v) => v.toString('base64'),
            deserialize: (v) => Buffer.from(v, 'base64'),
        }, 'Bytes');
    }
}
exports.registerCustomSerializers = registerCustomSerializers;
