// @zenstackhq/sdk/dmmf-helpers/modelArgs-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addMissingInputObjectTypesForModelArgs = void 0;
const model_helpers_1 = require("./model-helpers");
function addMissingInputObjectTypesForModelArgs(inputObjectTypes, models) {
    const modelArgsInputObjectTypes = generateModelArgsInputObjectTypes(models);
    for (const modelArgsInputObjectType of modelArgsInputObjectTypes) {
        inputObjectTypes.push(modelArgsInputObjectType);
    }
}
exports.addMissingInputObjectTypesForModelArgs = addMissingInputObjectTypesForModelArgs;
function generateModelArgsInputObjectTypes(models) {
    const modelArgsInputObjectTypes = [];
    for (const model of models) {
        const { name: modelName } = model;
        const fields = [];
        const selectField = {
            name: 'select',
            isRequired: false,
            isNullable: false,
            inputTypes: [
                {
                    isList: false,
                    type: `${modelName}Select`,
                    location: 'inputObjectTypes',
                    namespace: 'prisma',
                },
            ],
        };
        fields.push(selectField);
        const hasRelationToAnotherModel = (0, model_helpers_1.checkModelHasModelRelation)(model);
        if (hasRelationToAnotherModel) {
            const includeField = {
                name: 'include',
                isRequired: false,
                isNullable: false,
                inputTypes: [
                    {
                        isList: false,
                        type: `${modelName}Include`,
                        location: 'inputObjectTypes',
                        namespace: 'prisma',
                    },
                ],
            };
            fields.push(includeField);
        }
        const modelArgsInputObjectType = {
            name: `${modelName}Args`,
            constraints: {
                maxNumFields: null,
                minNumFields: null,
            },
            fields,
        };
        modelArgsInputObjectTypes.push(modelArgsInputObjectType);
    }
    return modelArgsInputObjectTypes;
}
