// ts-japi/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolymorphicSerializer = exports.Serializer = exports.ErrorSerializer = exports.JapiError = exports.Relator = exports.Paginator = exports.Metaizer = exports.Linker = exports.Cache = void 0;
var cache_1 = require("./classes/cache");
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return __importDefault(cache_1).default; } });
var linker_1 = require("./classes/linker");
Object.defineProperty(exports, "Linker", { enumerable: true, get: function () { return __importDefault(linker_1).default; } });
var metaizer_1 = require("./classes/metaizer");
Object.defineProperty(exports, "Metaizer", { enumerable: true, get: function () { return __importDefault(metaizer_1).default; } });
var paginator_1 = require("./classes/paginator");
Object.defineProperty(exports, "Paginator", { enumerable: true, get: function () { return __importDefault(paginator_1).default; } });
var relator_1 = require("./classes/relator");
Object.defineProperty(exports, "Relator", { enumerable: true, get: function () { return __importDefault(relator_1).default; } });
var error_model_1 = require("./models/error.model");
Object.defineProperty(exports, "JapiError", { enumerable: true, get: function () { return __importDefault(error_model_1).default; } });
var error_serializer_1 = require("./classes/error-serializer");
Object.defineProperty(exports, "ErrorSerializer", { enumerable: true, get: function () { return __importDefault(error_serializer_1).default; } });
var serializer_1 = require("./classes/serializer");
Object.defineProperty(exports, "Serializer", { enumerable: true, get: function () { return __importDefault(serializer_1).default; } });
var polymorphic_serialiser_1 = require("./classes/polymorphic-serialiser");
Object.defineProperty(exports, "PolymorphicSerializer", { enumerable: true, get: function () { return __importDefault(polymorphic_serialiser_1).default; } });
__exportStar(require("./interfaces/cache.interface"), exports);
__exportStar(require("./interfaces/error-serializer.interface"), exports);
__exportStar(require("./interfaces/error.interface"), exports);
__exportStar(require("./interfaces/json-api.interface"), exports);
__exportStar(require("./interfaces/linker.interface"), exports);
__exportStar(require("./interfaces/paginator.interface"), exports);
__exportStar(require("./interfaces/relator.interface"), exports);
__exportStar(require("./interfaces/serializer.interface"), exports);
__exportStar(require("./types/global.types"), exports);
__exportStar(require("./utils/is-error-document"), exports);
__exportStar(require("./utils/is-plain-object"), exports);
__exportStar(require("./utils/is-object"), exports);
