// langium/lib/lsp/workspace-symbol-provider.js
"use strict";
/******************************************************************************
 * Copyright 2023 TypeFox GmbH
 * This program and the accompanying materials are made available under the
 * terms of the MIT License, which is available in the project root.
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultWorkspaceSymbolProvider = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
const promise_util_1 = require("../utils/promise-util");
class DefaultWorkspaceSymbolProvider {
    constructor(services) {
        this.indexManager = services.workspace.IndexManager;
        this.nodeKindProvider = services.lsp.NodeKindProvider;
        this.fuzzyMatcher = services.lsp.FuzzyMatcher;
    }
    async getSymbols(params, cancelToken = vscode_languageserver_1.CancellationToken.None) {
        const workspaceSymbols = [];
        const query = params.query.toLowerCase();
        for (const description of this.indexManager.allElements()) {
            await (0, promise_util_1.interruptAndCheck)(cancelToken);
            if (this.fuzzyMatcher.match(query, description.name)) {
                const symbol = this.getWorkspaceSymbol(description);
                if (symbol) {
                    workspaceSymbols.push(symbol);
                }
            }
        }
        return workspaceSymbols;
    }
    getWorkspaceSymbol(astDescription) {
        const nameSegment = astDescription.nameSegment;
        if (nameSegment) {
            return {
                kind: this.nodeKindProvider.getSymbolKind(astDescription),
                name: astDescription.name,
                location: {
                    range: nameSegment.range,
                    uri: astDescription.documentUri.toString()
                }
            };
        }
        else {
            return undefined;
        }
    }
}
exports.DefaultWorkspaceSymbolProvider = DefaultWorkspaceSymbolProvider;
