// zenstack/plugins/zod/transformer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/ban-ts-comment */
const sdk_1 = require("@zenstackhq/sdk");
const dmmf_helpers_1 = require("@zenstackhq/sdk/dmmf-helpers");
const prisma_1 = require("@zenstackhq/sdk/prisma");
const path_1 = __importDefault(require("path"));
const upper_case_first_1 = require("upper-case-first");
const generator_1 = require("./generator");
class Transformer {
    constructor(params) {
        var _a, _b, _c, _d, _e, _f;
        this.schemaImports = new Set();
        this.hasJson = false;
        this.hasDecimal = false;
        this.sourceFiles = [];
        this.originalName = (_a = params.name) !== null && _a !== void 0 ? _a : '';
        this.name = params.name ? (0, upper_case_first_1.upperCaseFirst)(params.name) : '';
        this.fields = (_b = params.fields) !== null && _b !== void 0 ? _b : [];
        this.models = (_c = params.models) !== null && _c !== void 0 ? _c : [];
        this.modelOperations = (_d = params.modelOperations) !== null && _d !== void 0 ? _d : [];
        this.aggregateOperationSupport = (_e = params.aggregateOperationSupport) !== null && _e !== void 0 ? _e : {};
        this.enumTypes = (_f = params.enumTypes) !== null && _f !== void 0 ? _f : [];
        this.project = params.project;
        this.inputObjectTypes = params.inputObjectTypes;
        this.zmodel = params.zmodel;
    }
    static setOutputPath(outPath) {
        this.outputPath = outPath;
    }
    static getOutputPath() {
        return this.outputPath;
    }
    generateEnumSchemas() {
        return __awaiter(this, void 0, void 0, function* () {
            for (const enumType of this.enumTypes) {
                const name = (0, upper_case_first_1.upperCaseFirst)(enumType.name);
                const filePath = path_1.default.join(Transformer.outputPath, `enums/${name}.schema.ts`);
                const content = `/* eslint-disable */\n${this.generateImportZodStatement()}\n${this.generateExportSchemaStatement(`${name}`, `z.enum(${JSON.stringify(enumType.values)})`)}`;
                this.sourceFiles.push(this.project.createSourceFile(filePath, content, { overwrite: true }));
            }
            this.sourceFiles.push(this.project.createSourceFile(path_1.default.join(Transformer.outputPath, `enums/index.ts`), this.enumTypes
                .map((enumType) => `export * from './${(0, upper_case_first_1.upperCaseFirst)(enumType.name)}.schema';`)
                .join('\n'), { overwrite: true }));
        });
    }
    generateImportZodStatement() {
        return "import { z } from 'zod';\n";
    }
    generateExportSchemaStatement(name, schema) {
        return `export const ${name}Schema = ${schema}`;
    }
    generateObjectSchema(generateUnchecked, options) {
        const zodObjectSchemaFields = this.generateObjectSchemaFields(generateUnchecked);
        const objectSchema = this.prepareObjectSchema(zodObjectSchemaFields, options);
        const filePath = path_1.default.join(Transformer.outputPath, `objects/${this.name}.schema.ts`);
        const content = '/* eslint-disable */\n' + objectSchema;
        this.sourceFiles.push(this.project.createSourceFile(filePath, content, { overwrite: true }));
        return `${this.name}.schema`;
    }
    generateObjectSchemaFields(generateUnchecked) {
        const zodObjectSchemaFields = this.fields
            .map((field) => this.generateObjectSchemaField(field, generateUnchecked))
            .flatMap((item) => item)
            .map((item) => {
            const [zodStringWithMainType, field, skipValidators] = item;
            const value = skipValidators
                ? zodStringWithMainType
                : this.generateFieldValidators(zodStringWithMainType, field);
            return value.trim();
        });
        return zodObjectSchemaFields;
    }
    generateObjectSchemaField(field, generateUnchecked) {
        const lines = field.inputTypes;
        if (lines.length === 0) {
            return [];
        }
        let alternatives = lines.reduce((result, inputType) => {
            if (!generateUnchecked && typeof inputType.type === 'string' && inputType.type.includes('Unchecked')) {
                return result;
            }
            if (inputType.type.includes('CreateMany') && !(0, prisma_1.supportCreateMany)(this.zmodel)) {
                return result;
            }
            // TODO: unify the following with `schema-gen.ts`
            if (inputType.type === 'String') {
                result.push(this.wrapWithZodValidators('z.string()', field, inputType));
            }
            else if (inputType.type === 'Int' || inputType.type === 'Float') {
                result.push(this.wrapWithZodValidators('z.number()', field, inputType));
            }
            else if (inputType.type === 'Decimal') {
                this.hasDecimal = true;
                result.push(this.wrapWithZodValidators('DecimalSchema', field, inputType));
            }
            else if (inputType.type === 'BigInt') {
                result.push(this.wrapWithZodValidators('z.bigint()', field, inputType));
            }
            else if (inputType.type === 'Boolean') {
                result.push(this.wrapWithZodValidators('z.boolean()', field, inputType));
            }
            else if (inputType.type === 'DateTime') {
                result.push(this.wrapWithZodValidators(['z.date()', 'z.string().datetime()'], field, inputType));
            }
            else if (inputType.type === 'Bytes') {
                result.push(this.wrapWithZodValidators(`z.custom<Buffer | Uint8Array>(data => data instanceof Uint8Array)`, field, inputType));
            }
            else if (inputType.type === 'Json') {
                this.hasJson = true;
                result.push(this.wrapWithZodValidators('jsonSchema', field, inputType));
            }
            else if (inputType.type === 'True') {
                result.push(this.wrapWithZodValidators('z.literal(true)', field, inputType));
            }
            else if (inputType.type === 'Null') {
                result.push(this.wrapWithZodValidators('z.null()', field, inputType));
            }
            else {
                const isEnum = inputType.location === 'enumTypes';
                const isFieldRef = inputType.location === 'fieldRefTypes';
                if (
                // fieldRefTypes refer to other fields in the model and don't need to be generated as part of schema
                !isFieldRef &&
                    (inputType.namespace === 'prisma' || isEnum)) {
                    if (inputType.type !== this.originalName && typeof inputType.type === 'string') {
                        this.addSchemaImport(inputType.type);
                    }
                    result.push(this.generatePrismaStringLine(field, inputType, lines.length));
                }
            }
            return result;
        }, []);
        if (alternatives.length === 0) {
            return [];
        }
        if (alternatives.length > 1) {
            alternatives = alternatives.map((alter) => alter.replace('.optional()', ''));
        }
        const fieldName = alternatives.some((alt) => alt.includes(':')) ? '' : `  ${field.name}:`;
        const opt = !field.isRequired ? '.optional()' : '';
        let resString = alternatives.length === 1 ? alternatives.join(',\r\n') : `z.union([${alternatives.join(',\r\n')}])${opt}`;
        if (field.isNullable) {
            resString += '.nullable()';
        }
        return [[`  ${fieldName} ${resString} `, field, true]];
    }
    wrapWithZodValidators(mainValidators, field, inputType) {
        let line = '';
        const base = Array.isArray(mainValidators) ? mainValidators : [mainValidators];
        line += base
            .map((validator) => {
            let r = validator;
            if (inputType.isList) {
                r += '.array()';
            }
            if (!field.isRequired) {
                r += '.optional()';
            }
            return r;
        })
            .join(', ');
        if (base.length > 1) {
            line = `z.union([${line}])`;
        }
        return line;
    }
    addSchemaImport(name) {
        this.schemaImports.add((0, upper_case_first_1.upperCaseFirst)(name));
    }
    generatePrismaStringLine(field, inputType, inputsLength) {
        const isEnum = inputType.location === 'enumTypes';
        const { isModelQueryType, modelName, queryName } = this.checkIsModelQueryType(inputType.type);
        const objectSchemaLine = isModelQueryType
            ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                this.resolveModelQuerySchemaName(modelName, queryName)
            : `${(0, upper_case_first_1.upperCaseFirst)(inputType.type.toString())}ObjectSchema`;
        const enumSchemaLine = `${(0, upper_case_first_1.upperCaseFirst)(inputType.type.toString())}Schema`;
        const schema = inputType.type === this.name ? objectSchemaLine : isEnum ? enumSchemaLine : objectSchemaLine;
        const arr = inputType.isList ? '.array()' : '';
        const opt = !field.isRequired ? '.optional()' : '';
        return inputsLength === 1
            ? `  ${field.name}: z.lazy(() => ${schema})${arr}${opt}`
            : `z.lazy(() => ${schema})${arr}${opt}`;
    }
    generateFieldValidators(zodStringWithMainType, field) {
        const { isRequired, isNullable } = field;
        if (!isRequired) {
            zodStringWithMainType += '.optional()';
        }
        if (isNullable) {
            zodStringWithMainType += '.nullable()';
        }
        return zodStringWithMainType;
    }
    prepareObjectSchema(zodObjectSchemaFields, options) {
        const objectSchema = `${this.generateExportObjectSchemaStatement(this.addFinalWrappers({ zodStringFields: zodObjectSchemaFields }))}\n`;
        const prismaImportStatement = this.generateImportPrismaStatement(options);
        const json = this.generateJsonSchemaImplementation();
        return `${this.generateObjectSchemaImportStatements()}${prismaImportStatement}${json}${objectSchema}`;
    }
    generateExportObjectSchemaStatement(schema) {
        let name = this.name;
        let origName = this.originalName;
        if ((0, dmmf_helpers_1.isAggregateInputType)(name)) {
            name = `${name}Type`;
            origName = `${origName}Type`;
        }
        const outType = `z.ZodType<Prisma.${origName}>`;
        return `type SchemaType = ${outType};
export const ${this.name}ObjectSchema: SchemaType = ${schema} as SchemaType;`;
    }
    addFinalWrappers({ zodStringFields }) {
        const fields = [...zodStringFields];
        return this.wrapWithZodObject(fields) + '.strict()';
    }
    generateImportPrismaStatement(options) {
        const prismaClientImportPath = (0, generator_1.computePrismaClientImport)(path_1.default.resolve(Transformer.outputPath, './objects'), options);
        return `import type { Prisma } from '${prismaClientImportPath}';\n\n`;
    }
    generateJsonSchemaImplementation() {
        let jsonSchemaImplementation = '';
        if (this.hasJson) {
            jsonSchemaImplementation += `\n`;
            jsonSchemaImplementation += `const literalSchema = z.union([z.string(), z.number(), z.boolean()]);\n`;
            jsonSchemaImplementation += `const jsonSchema: z.ZodType<Prisma.InputJsonValue> = z.lazy(() =>\n`;
            jsonSchemaImplementation += `  z.union([literalSchema, z.array(jsonSchema.nullable()), z.record(jsonSchema.nullable())])\n`;
            jsonSchemaImplementation += `);\n\n`;
        }
        return jsonSchemaImplementation;
    }
    generateObjectSchemaImportStatements() {
        let generatedImports = this.generateImportZodStatement();
        generatedImports += this.generateSchemaImports();
        generatedImports += this.generateCommonImport();
        generatedImports += '\n\n';
        return generatedImports;
    }
    generateSchemaImports() {
        return [...this.schemaImports]
            .map((name) => {
            const { isModelQueryType, modelName } = this.checkIsModelQueryType(name);
            if (isModelQueryType) {
                return `import { ${modelName}InputSchema } from '../input/${modelName}Input.schema';`;
            }
            else if (Transformer.enumNames.includes(name)) {
                return `import { ${name}Schema } from '../enums/${name}.schema';`;
            }
            else {
                return `import { ${name}ObjectSchema } from './${name}.schema';`;
            }
        })
            .join('\n');
    }
    generateCommonImport() {
        let r = '';
        if (this.hasDecimal) {
            r += `import { DecimalSchema } from '../common';\n`;
        }
        if (r) {
            r += '\n';
        }
        return r;
    }
    checkIsModelQueryType(type) {
        const modelQueryTypeSuffixToQueryName = {
            FindManyArgs: 'findMany',
        };
        for (const modelQueryType of ['FindManyArgs']) {
            if (type.includes(modelQueryType)) {
                const modelQueryTypeSuffixIndex = type.indexOf(modelQueryType);
                return {
                    isModelQueryType: true,
                    modelName: (0, upper_case_first_1.upperCaseFirst)(type.substring(0, modelQueryTypeSuffixIndex)),
                    queryName: modelQueryTypeSuffixToQueryName[modelQueryType],
                };
            }
        }
        return { isModelQueryType: false };
    }
    resolveModelQuerySchemaName(modelName, queryName) {
        return `${modelName}InputSchema.${queryName}`;
    }
    wrapWithZodUnion(zodStringFields) {
        let wrapped = '';
        wrapped += 'z.union([';
        wrapped += '\n';
        wrapped += '  ' + zodStringFields.join(',');
        wrapped += '\n';
        wrapped += '])';
        return wrapped;
    }
    wrapWithZodObject(zodStringFields) {
        let wrapped = '';
        wrapped += 'z.object({';
        wrapped += '\n';
        wrapped += '  ' + zodStringFields;
        wrapped += '\n';
        wrapped += '})';
        return wrapped;
    }
    generateInputSchemas(options, zmodel) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e;
            const globalExports = [];
            // whether Prisma's Unchecked* series of input types should be generated
            const generateUnchecked = options.noUncheckedInput !== true;
            for (const modelOperation of this.modelOperations) {
                const { model: origModelName, findUnique, findFirst, findMany, 
                // @ts-expect-error
                createOne, createMany, 
                // @ts-expect-error
                deleteOne, 
                // @ts-expect-error
                updateOne, deleteMany, updateMany, 
                // @ts-expect-error
                upsertOne, aggregate, groupBy,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                 } = modelOperation;
                const modelName = (0, upper_case_first_1.upperCaseFirst)(origModelName);
                globalExports.push(`export * from './${modelName}Input.schema'`);
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const model = (0, dmmf_helpers_1.findModelByName)(this.models, origModelName);
                const { selectImport, includeImport, selectZodSchemaLineLazy, includeZodSchemaLineLazy } = this.resolveSelectIncludeImportAndZodSchemaLine(model);
                let imports = [
                    `import { z } from 'zod'`,
                    this.generateImportPrismaStatement(options),
                    selectImport,
                    includeImport,
                ];
                let codeBody = '';
                const operations = [];
                // OrderByWithRelationInput's name is different when "fullTextSearch" is enabled
                const orderByWithRelationInput = this.inputObjectTypes
                    .map((o) => (0, upper_case_first_1.upperCaseFirst)(o.name))
                    .includes(`${modelName}OrderByWithRelationInput`)
                    ? `${modelName}OrderByWithRelationInput`
                    : `${modelName}OrderByWithRelationAndSearchRelevanceInput`;
                if (findUnique) {
                    imports.push(`import { ${modelName}WhereUniqueInputObjectSchema } from '../objects/${modelName}WhereUniqueInput.schema'`);
                    codeBody += `findUnique: z.object({ ${selectZodSchemaLineLazy} ${includeZodSchemaLineLazy} where: ${modelName}WhereUniqueInputObjectSchema }),`;
                    operations.push(['findUnique', origModelName]);
                }
                if (findFirst) {
                    imports.push(`import { ${modelName}WhereInputObjectSchema } from '../objects/${modelName}WhereInput.schema'`, `import { ${orderByWithRelationInput}ObjectSchema } from '../objects/${orderByWithRelationInput}.schema'`, `import { ${modelName}WhereUniqueInputObjectSchema } from '../objects/${modelName}WhereUniqueInput.schema'`, `import { ${modelName}ScalarFieldEnumSchema } from '../enums/${modelName}ScalarFieldEnum.schema'`);
                    codeBody += `findFirst: z.object({ ${selectZodSchemaLineLazy} ${includeZodSchemaLineLazy} where: ${modelName}WhereInputObjectSchema.optional(), orderBy: z.union([${orderByWithRelationInput}ObjectSchema, ${orderByWithRelationInput}ObjectSchema.array()]).optional(), cursor: ${modelName}WhereUniqueInputObjectSchema.optional(), take: z.number().optional(), skip: z.number().optional(), distinct: z.array(${modelName}ScalarFieldEnumSchema).optional() }),`;
                    operations.push(['findFirst', origModelName]);
                }
                if (findMany) {
                    imports.push(`import { ${modelName}WhereInputObjectSchema } from '../objects/${modelName}WhereInput.schema'`, `import { ${orderByWithRelationInput}ObjectSchema } from '../objects/${orderByWithRelationInput}.schema'`, `import { ${modelName}WhereUniqueInputObjectSchema } from '../objects/${modelName}WhereUniqueInput.schema'`, `import { ${modelName}ScalarFieldEnumSchema } from '../enums/${modelName}ScalarFieldEnum.schema'`);
                    codeBody += `findMany: z.object({ ${selectZodSchemaLineLazy} ${includeZodSchemaLineLazy} where: ${modelName}WhereInputObjectSchema.optional(), orderBy: z.union([${orderByWithRelationInput}ObjectSchema, ${orderByWithRelationInput}ObjectSchema.array()]).optional(), cursor: ${modelName}WhereUniqueInputObjectSchema.optional(), take: z.number().optional(), skip: z.number().optional(), distinct: z.array(${modelName}ScalarFieldEnumSchema).optional()  }),`;
                    operations.push(['findMany', origModelName]);
                }
                if (createOne) {
                    imports.push(`import { ${modelName}CreateInputObjectSchema } from '../objects/${modelName}CreateInput.schema'`);
                    if (generateUnchecked) {
                        imports.push(`import { ${modelName}UncheckedCreateInputObjectSchema } from '../objects/${modelName}UncheckedCreateInput.schema'`);
                    }
                    const dataSchema = generateUnchecked
                        ? `z.union([${modelName}CreateInputObjectSchema, ${modelName}UncheckedCreateInputObjectSchema])`
                        : `${modelName}CreateInputObjectSchema`;
                    codeBody += `create: z.object({ ${selectZodSchemaLineLazy} ${includeZodSchemaLineLazy} data: ${dataSchema} }),`;
                    operations.push(['create', origModelName]);
                }
                if (createMany && (0, prisma_1.supportCreateMany)(zmodel)) {
                    imports.push(`import { ${modelName}CreateManyInputObjectSchema } from '../objects/${modelName}CreateManyInput.schema'`);
                    codeBody += `createMany: z.object({ data: z.union([${modelName}CreateManyInputObjectSchema, z.array(${modelName}CreateManyInputObjectSchema)]), skipDuplicates: z.boolean().optional() }),`;
                    operations.push(['createMany', origModelName]);
                }
                if (deleteOne) {
                    imports.push(`import { ${modelName}WhereUniqueInputObjectSchema } from '../objects/${modelName}WhereUniqueInput.schema'`);
                    codeBody += `'delete': z.object({ ${selectZodSchemaLineLazy} ${includeZodSchemaLineLazy} where: ${modelName}WhereUniqueInputObjectSchema  }),`;
                    operations.push(['delete', origModelName]);
                }
                if (deleteMany) {
                    imports.push(`import { ${modelName}WhereInputObjectSchema } from '../objects/${modelName}WhereInput.schema'`);
                    codeBody += `deleteMany: z.object({ where: ${modelName}WhereInputObjectSchema.optional()  }),`;
                    operations.push(['deleteMany', origModelName]);
                }
                if (updateOne) {
                    imports.push(`import { ${modelName}UpdateInputObjectSchema } from '../objects/${modelName}UpdateInput.schema'`, `import { ${modelName}WhereUniqueInputObjectSchema } from '../objects/${modelName}WhereUniqueInput.schema'`);
                    if (generateUnchecked) {
                        imports.push(`import { ${modelName}UncheckedUpdateInputObjectSchema } from '../objects/${modelName}UncheckedUpdateInput.schema'`);
                    }
                    const dataSchema = generateUnchecked
                        ? `z.union([${modelName}UpdateInputObjectSchema, ${modelName}UncheckedUpdateInputObjectSchema])`
                        : `${modelName}UpdateInputObjectSchema`;
                    codeBody += `update: z.object({ ${selectZodSchemaLineLazy} ${includeZodSchemaLineLazy} data: ${dataSchema}, where: ${modelName}WhereUniqueInputObjectSchema  }),`;
                    operations.push(['update', origModelName]);
                }
                if (updateMany) {
                    imports.push(`import { ${modelName}UpdateManyMutationInputObjectSchema } from '../objects/${modelName}UpdateManyMutationInput.schema'`, `import { ${modelName}WhereInputObjectSchema } from '../objects/${modelName}WhereInput.schema'`);
                    if (generateUnchecked) {
                        imports.push(`import { ${modelName}UncheckedUpdateManyInputObjectSchema } from '../objects/${modelName}UncheckedUpdateManyInput.schema'`);
                    }
                    const dataSchema = generateUnchecked
                        ? `z.union([${modelName}UpdateManyMutationInputObjectSchema, ${modelName}UncheckedUpdateManyInputObjectSchema])`
                        : `${modelName}UpdateManyMutationInputObjectSchema`;
                    codeBody += `updateMany: z.object({ data: ${dataSchema}, where: ${modelName}WhereInputObjectSchema.optional()  }),`;
                    operations.push(['updateMany', origModelName]);
                }
                if (upsertOne) {
                    imports.push(`import { ${modelName}WhereUniqueInputObjectSchema } from '../objects/${modelName}WhereUniqueInput.schema'`, `import { ${modelName}CreateInputObjectSchema } from '../objects/${modelName}CreateInput.schema'`, `import { ${modelName}UpdateInputObjectSchema } from '../objects/${modelName}UpdateInput.schema'`);
                    if (generateUnchecked) {
                        imports.push(`import { ${modelName}UncheckedCreateInputObjectSchema } from '../objects/${modelName}UncheckedCreateInput.schema'`, `import { ${modelName}UncheckedUpdateInputObjectSchema } from '../objects/${modelName}UncheckedUpdateInput.schema'`);
                    }
                    const createSchema = generateUnchecked
                        ? `z.union([${modelName}CreateInputObjectSchema, ${modelName}UncheckedCreateInputObjectSchema])`
                        : `${modelName}CreateInputObjectSchema`;
                    const updateSchema = generateUnchecked
                        ? `z.union([${modelName}UpdateInputObjectSchema, ${modelName}UncheckedUpdateInputObjectSchema])`
                        : `${modelName}UpdateInputObjectSchema`;
                    codeBody += `upsert: z.object({ ${selectZodSchemaLineLazy} ${includeZodSchemaLineLazy} where: ${modelName}WhereUniqueInputObjectSchema, create: ${createSchema}, update: ${updateSchema} }),`;
                    operations.push(['upsert', origModelName]);
                }
                const aggregateOperations = [];
                if ((_a = this.aggregateOperationSupport[modelName]) === null || _a === void 0 ? void 0 : _a.count) {
                    imports.push(`import { ${modelName}CountAggregateInputObjectSchema } from '../objects/${modelName}CountAggregateInput.schema'`);
                    aggregateOperations.push(`_count: z.union([ z.literal(true), ${modelName}CountAggregateInputObjectSchema ]).optional()`);
                }
                if ((_b = this.aggregateOperationSupport[modelName]) === null || _b === void 0 ? void 0 : _b.min) {
                    imports.push(`import { ${modelName}MinAggregateInputObjectSchema } from '../objects/${modelName}MinAggregateInput.schema'`);
                    aggregateOperations.push(`_min: ${modelName}MinAggregateInputObjectSchema.optional()`);
                }
                if ((_c = this.aggregateOperationSupport[modelName]) === null || _c === void 0 ? void 0 : _c.max) {
                    imports.push(`import { ${modelName}MaxAggregateInputObjectSchema } from '../objects/${modelName}MaxAggregateInput.schema'`);
                    aggregateOperations.push(`_max: ${modelName}MaxAggregateInputObjectSchema.optional()`);
                }
                if ((_d = this.aggregateOperationSupport[modelName]) === null || _d === void 0 ? void 0 : _d.avg) {
                    imports.push(`import { ${modelName}AvgAggregateInputObjectSchema } from '../objects/${modelName}AvgAggregateInput.schema'`);
                    aggregateOperations.push(`_avg: ${modelName}AvgAggregateInputObjectSchema.optional()`);
                }
                if ((_e = this.aggregateOperationSupport[modelName]) === null || _e === void 0 ? void 0 : _e.sum) {
                    imports.push(`import { ${modelName}SumAggregateInputObjectSchema } from '../objects/${modelName}SumAggregateInput.schema'`);
                    aggregateOperations.push(`_sum: ${modelName}SumAggregateInputObjectSchema.optional()`);
                }
                if (aggregate) {
                    imports.push(`import { ${modelName}WhereInputObjectSchema } from '../objects/${modelName}WhereInput.schema'`, `import { ${orderByWithRelationInput}ObjectSchema } from '../objects/${orderByWithRelationInput}.schema'`, `import { ${modelName}WhereUniqueInputObjectSchema } from '../objects/${modelName}WhereUniqueInput.schema'`);
                    codeBody += `aggregate: z.object({ where: ${modelName}WhereInputObjectSchema.optional(), orderBy: z.union([${orderByWithRelationInput}ObjectSchema, ${orderByWithRelationInput}ObjectSchema.array()]).optional(), cursor: ${modelName}WhereUniqueInputObjectSchema.optional(), take: z.number().optional(), skip: z.number().optional(), ${aggregateOperations.join(', ')} }),`;
                    operations.push(['aggregate', modelName]);
                }
                if (groupBy) {
                    imports.push(`import { ${modelName}WhereInputObjectSchema } from '../objects/${modelName}WhereInput.schema'`, `import { ${modelName}OrderByWithAggregationInputObjectSchema } from '../objects/${modelName}OrderByWithAggregationInput.schema'`, `import { ${modelName}ScalarWhereWithAggregatesInputObjectSchema } from '../objects/${modelName}ScalarWhereWithAggregatesInput.schema'`, `import { ${modelName}ScalarFieldEnumSchema } from '../enums/${modelName}ScalarFieldEnum.schema'`);
                    codeBody += `groupBy: z.object({ where: ${modelName}WhereInputObjectSchema.optional(), orderBy: z.union([${modelName}OrderByWithAggregationInputObjectSchema, ${modelName}OrderByWithAggregationInputObjectSchema.array()]).optional(), having: ${modelName}ScalarWhereWithAggregatesInputObjectSchema.optional(), take: z.number().optional(), skip: z.number().optional(), by: z.array(${modelName}ScalarFieldEnumSchema), ${aggregateOperations.join(', ')} }),`;
                    operations.push(['groupBy', origModelName]);
                }
                // count
                {
                    imports.push(`import { ${modelName}WhereInputObjectSchema } from '../objects/${modelName}WhereInput.schema'`, `import { ${orderByWithRelationInput}ObjectSchema } from '../objects/${orderByWithRelationInput}.schema'`, `import { ${modelName}WhereUniqueInputObjectSchema } from '../objects/${modelName}WhereUniqueInput.schema'`, `import { ${modelName}ScalarFieldEnumSchema } from '../enums/${modelName}ScalarFieldEnum.schema'`, `import { ${modelName}CountAggregateInputObjectSchema } from '../objects/${modelName}CountAggregateInput.schema'`);
                    codeBody += `count: z.object({ where: ${modelName}WhereInputObjectSchema.optional(), orderBy: z.union([${orderByWithRelationInput}ObjectSchema, ${orderByWithRelationInput}ObjectSchema.array()]).optional(), cursor: ${modelName}WhereUniqueInputObjectSchema.optional(), take: z.number().optional(), skip: z.number().optional(), distinct: z.array(${modelName}ScalarFieldEnumSchema).optional(), select: z.union([ z.literal(true), ${modelName}CountAggregateInputObjectSchema ]).optional() })`;
                    operations.push(['count', origModelName]);
                }
                imports = [...new Set(imports)];
                const filePath = path_1.default.join(Transformer.outputPath, `input/${modelName}Input.schema.ts`);
                const content = `
            /* eslint-disable */
            ${imports.join(';\n')}
            
            type ${modelName}InputSchemaType = {
${operations
                    .map(([operation, typeName]) => (0, sdk_1.indentString)(`${operation}: z.ZodType<Prisma.${typeName}${(0, upper_case_first_1.upperCaseFirst)(operation)}Args>`, 4))
                    .join(',\n')}
            }

            export const ${modelName}InputSchema = {
            ${(0, sdk_1.indentString)(codeBody, 4)}
            } as ${modelName}InputSchemaType;
                        `;
                this.sourceFiles.push(this.project.createSourceFile(filePath, content, { overwrite: true }));
            }
            const indexFilePath = path_1.default.join(Transformer.outputPath, 'input/index.ts');
            const indexContent = `
/* eslint-disable */
${globalExports.join(';\n')}
`;
            this.sourceFiles.push(this.project.createSourceFile(indexFilePath, indexContent, { overwrite: true }));
        });
    }
    generateImportStatements(imports) {
        var _a;
        let generatedImports = this.generateImportZodStatement();
        generatedImports += (_a = imports === null || imports === void 0 ? void 0 : imports.filter((importItem) => !!importItem).join(';\r\n')) !== null && _a !== void 0 ? _a : '';
        generatedImports += '\n\n';
        return generatedImports;
    }
    resolveSelectIncludeImportAndZodSchemaLine(model) {
        const { name } = model;
        const modelName = (0, upper_case_first_1.upperCaseFirst)(name);
        const hasRelationToAnotherModel = (0, dmmf_helpers_1.checkModelHasModelRelation)(model);
        const selectImport = `import { ${modelName}SelectObjectSchema } from '../objects/${modelName}Select.schema'`;
        const includeImport = hasRelationToAnotherModel
            ? `import { ${modelName}IncludeObjectSchema } from '../objects/${modelName}Include.schema'`
            : '';
        let selectZodSchemaLine = '';
        let includeZodSchemaLine = '';
        let selectZodSchemaLineLazy = '';
        let includeZodSchemaLineLazy = '';
        const zodSelectObjectSchema = `${modelName}SelectObjectSchema.optional()`;
        selectZodSchemaLine = `select: ${zodSelectObjectSchema},`;
        selectZodSchemaLineLazy = `select: z.lazy(() => ${zodSelectObjectSchema}),`;
        if (hasRelationToAnotherModel) {
            const zodIncludeObjectSchema = `${modelName}IncludeObjectSchema.optional()`;
            includeZodSchemaLine = `include: ${zodIncludeObjectSchema},`;
            includeZodSchemaLineLazy = `include: z.lazy(() => ${zodIncludeObjectSchema}),`;
        }
        return {
            selectImport,
            includeImport,
            selectZodSchemaLine,
            includeZodSchemaLine,
            selectZodSchemaLineLazy,
            includeZodSchemaLineLazy,
        };
    }
}
Transformer.enumNames = [];
Transformer.rawOpsMap = {};
Transformer.outputPath = './generated';
exports.default = Transformer;
