// @prisma/internals/dist/chunk-WGS7QFFN.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chunk_WGS7QFFN_exports = {};
__export(chunk_WGS7QFFN_exports, {
  ValidateError: () => ValidateError,
  validate: () => validate
});
module.exports = __toCommonJS(chunk_WGS7QFFN_exports);
var import_chunk_URGOXRQA = require("./chunk-URGOXRQA.js");
var import_chunk_G2BFX4VJ = require("./chunk-G2BFX4VJ.js");
var import_chunk_WMIBAVWZ = require("./chunk-WMIBAVWZ.js");
var import_chunk_HO25CWRL = require("./chunk-HO25CWRL.js");
var import_chunk_NZ262BVM = require("./chunk-NZ262BVM.js");
var import_chunk_2IKVYFVK = require("./chunk-2IKVYFVK.js");
var import_chunk_6QRPFUVD = require("./chunk-6QRPFUVD.js");
var import_chunk_IKGYNTE3 = require("./chunk-IKGYNTE3.js");
var import_chunk_A4QWQE3W = require("./chunk-A4QWQE3W.js");
var import_chunk_PG5FDKSF = require("./chunk-PG5FDKSF.js");
var import_chunk_CTBG5EFY = require("./chunk-CTBG5EFY.js");
var import_debug = __toESM(require("@prisma/debug"));
var import_function = (0, import_chunk_CTBG5EFY.__toESM)((0, import_chunk_NZ262BVM.require_function)());
var debug = (0, import_debug.default)("prisma:validate");
var ValidateError = class extends Error {
  constructor(params) {
    const constructedErrorMessage = (0, import_chunk_IKGYNTE3.$)(params).with({ _tag: "parsed" }, ({ errorCode, message, reason }) => {
      const errorCodeMessage = errorCode ? `Error code: ${errorCode}` : "";
      return `${reason}
${errorCodeMessage}
${(0, import_chunk_WMIBAVWZ.relativizePathInPSLError)(message)}`;
    }).with({ _tag: "unparsed" }, ({ message, reason }) => {
      const detailsHeader = (0, import_chunk_PG5FDKSF.red)((0, import_chunk_PG5FDKSF.bold)("Details:"));
      return `${reason}
${detailsHeader} ${message}`;
    }).exhaustive();
    const errorMessageWithContext = `${constructedErrorMessage}
[Context: validate]`;
    super((0, import_chunk_HO25CWRL.addVersionDetailsToErrorMessage)(errorMessageWithContext));
    this.name = "ValidateError";
  }
};
function validate(options) {
  const debugErrorType = (0, import_chunk_G2BFX4VJ.createDebugErrorType)(debug, "validateWasm");
  debug(`Using validate Wasm`);
  const validateEither = (0, import_function.pipe)(
    (0, import_chunk_6QRPFUVD.tryCatch)(
      () => {
        if (process.env.FORCE_PANIC_QUERY_ENGINE_GET_DMMF) {
          debug("Triggering a Rust panic...");
          import_chunk_2IKVYFVK.prismaSchemaWasm.debug_panic();
        }
        const params = JSON.stringify({
          prismaSchema: options.schemas,
          noColor: Boolean(process.env.NO_COLOR)
        });
        import_chunk_2IKVYFVK.prismaSchemaWasm.validate(params);
      },
      (e) => ({
        type: "wasm-error",
        reason: "(validate wasm)",
        error: e
      })
    )
  );
  if ((0, import_chunk_6QRPFUVD.isRight)(validateEither)) {
    return;
  }
  const error = (0, import_chunk_IKGYNTE3.$)(validateEither.left).with({ type: "wasm-error" }, (e) => {
    debugErrorType(e);
    console.error("");
    if ((0, import_chunk_A4QWQE3W.isWasmPanic)(e.error)) {
      const { message, stack } = (0, import_chunk_A4QWQE3W.getWasmError)(e.error);
      debug(`Error validating schema: ${message}`);
      debug(stack);
      const panic = new import_chunk_A4QWQE3W.RustPanic(
        /* message */
        message,
        /* rustStack */
        stack,
        /* request */
        "@prisma/prisma-schema-wasm validate",
        "FMT_CLI",
        /* schemaPath */
        (0, import_chunk_URGOXRQA.debugMultipleSchemaPaths)(options.schemas),
        /* schema */
        (0, import_chunk_URGOXRQA.debugMultipleSchemas)(options.schemas)
      );
      return panic;
    }
    const errorOutput = e.error.message;
    return new ValidateError((0, import_chunk_G2BFX4VJ.parseQueryEngineError)({ errorOutput, reason: e.reason }));
  }).exhaustive();
  throw error;
}
