// @prisma/internals/dist/chunk-CKNTB45N.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chunk_CKNTB45N_exports = {};
__export(chunk_CKNTB45N_exports, {
  ClientEngineType: () => ClientEngineType,
  DEFAULT_CLIENT_ENGINE_TYPE: () => DEFAULT_CLIENT_ENGINE_TYPE,
  getClientEngineType: () => getClientEngineType
});
module.exports = __toCommonJS(chunk_CKNTB45N_exports);
var ClientEngineType = /* @__PURE__ */ ((ClientEngineType2) => {
  ClientEngineType2["Library"] = "library";
  ClientEngineType2["Binary"] = "binary";
  return ClientEngineType2;
})(ClientEngineType || {});
var DEFAULT_CLIENT_ENGINE_TYPE = "library";
function getClientEngineType(generatorConfig) {
  const engineTypeFromEnvVar = getEngineTypeFromEnvVar();
  if (engineTypeFromEnvVar) return engineTypeFromEnvVar;
  if (generatorConfig?.config.engineType === "library") {
    return "library";
  } else if (generatorConfig?.config.engineType === "binary") {
    return "binary";
  } else {
    return DEFAULT_CLIENT_ENGINE_TYPE;
  }
}
function getEngineTypeFromEnvVar() {
  const engineType = process.env.PRISMA_CLIENT_ENGINE_TYPE;
  if (engineType === "library") {
    return "library";
  } else if (engineType === "binary") {
    return "binary";
  } else {
    return void 0;
  }
}
