// @prisma/internals/dist/chunk-3JWDPPF7.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chunk_3JWDPPF7_exports = {};
__export(chunk_3JWDPPF7_exports, {
  handlePanic: () => handlePanic
});
module.exports = __toCommonJS(chunk_3JWDPPF7_exports);
var import_chunk_LNSJVWED = require("./chunk-LNSJVWED.js");
var import_chunk_UCTEZFJP = require("./chunk-UCTEZFJP.js");
var import_chunk_EPJYKNCQ = require("./chunk-EPJYKNCQ.js");
var import_chunk_27F53W4Q = require("./chunk-27F53W4Q.js");
var import_chunk_PG5FDKSF = require("./chunk-PG5FDKSF.js");
var import_prompts = __toESM(require("prompts"));
async function handlePanic(args) {
  if (!(0, import_chunk_EPJYKNCQ.canPrompt)()) {
    throw args.error;
  }
  await panicDialog(args);
}
async function panicDialog({
  error,
  cliVersion,
  enginesVersion,
  command,
  getDatabaseVersionSafe
}) {
  const errorMessage = error.message.split("\n").slice(0, Math.max(20, process.stdout.rows)).join("\n");
  console.log(`${(0, import_chunk_PG5FDKSF.red)("Oops, an unexpected error occurred!")}
${(0, import_chunk_PG5FDKSF.red)(errorMessage)}

${(0, import_chunk_PG5FDKSF.bold)("Please help us improve Prisma by submitting an error report.")}
${(0, import_chunk_PG5FDKSF.bold)("Error reports never contain personal or other sensitive information.")}
${(0, import_chunk_PG5FDKSF.dim)(`Learn more: ${(0, import_chunk_LNSJVWED.link)("https://pris.ly/d/telemetry")}`)}
`);
  const { value: shouldSubmitReport } = await (0, import_prompts.default)({
    type: "select",
    name: "value",
    message: "Submit error report",
    initial: 0,
    choices: [
      {
        title: "Yes",
        value: true,
        description: `Send error report once`
      },
      {
        title: "No",
        value: false,
        description: `Don't send error report`
      }
    ]
  });
  if (shouldSubmitReport) {
    try {
      console.log("Submitting...");
      const reportId = await (0, import_chunk_27F53W4Q.sendPanic)({ error, cliVersion, enginesVersion, getDatabaseVersionSafe });
      console.log(`
${(0, import_chunk_PG5FDKSF.bold)(`We successfully received the error report id: ${reportId}`)}`);
      console.log(`
${(0, import_chunk_PG5FDKSF.bold)("Thanks a lot for your help! \u{1F64F}")}`);
    } catch (error2) {
      const reportFailedMessage = `${(0, import_chunk_PG5FDKSF.bold)((0, import_chunk_PG5FDKSF.red)("Oops. We could not send the error report."))}`;
      console.log(reportFailedMessage);
      console.error(`${(0, import_chunk_PG5FDKSF.gray)("Error report submission failed due to: ")}`, error2);
    }
  }
  await (0, import_chunk_UCTEZFJP.wouldYouLikeToCreateANewIssue)({
    prompt: !shouldSubmitReport,
    error,
    cliVersion,
    enginesVersion,
    command
  });
  process.exit(1);
}
