// @prisma/internals/dist/logger.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var logger_exports = {};
__export(logger_exports, {
  error: () => import_chunk_TOTPZIKH.error,
  info: () => import_chunk_TOTPZIKH.info,
  log: () => import_chunk_TOTPZIKH.log,
  query: () => import_chunk_TOTPZIKH.query,
  should: () => import_chunk_TOTPZIKH.should,
  tags: () => import_chunk_TOTPZIKH.tags,
  warn: () => import_chunk_TOTPZIKH.warn
});
module.exports = __toCommonJS(logger_exports);
var import_chunk_TOTPZIKH = require("./chunk-TOTPZIKH.js");
var import_chunk_PG5FDKSF = require("./chunk-PG5FDKSF.js");
var import_chunk_CTBG5EFY = require("./chunk-CTBG5EFY.js");
