// @prisma/internals/dist/utils/fs-utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var fs_utils_exports = {};
__export(fs_utils_exports, {
  createDirIfNotExists: () => import_chunk_E7ECASRK.createDirIfNotExists,
  getFilesInDir: () => import_chunk_E7ECASRK.getFilesInDir,
  getNestedFoldersInDir: () => import_chunk_E7ECASRK.getNestedFoldersInDir,
  getTopLevelFoldersInDir: () => import_chunk_E7ECASRK.getTopLevelFoldersInDir,
  removeEmptyDirs: () => import_chunk_E7ECASRK.removeEmptyDirs,
  writeFile: () => import_chunk_E7ECASRK.writeFile
});
module.exports = __toCommonJS(fs_utils_exports);
var import_chunk_E7ECASRK = require("../chunk-E7ECASRK.js");
var import_chunk_J77ZIDXB = require("../chunk-J77ZIDXB.js");
var import_chunk_YHCFGHQV = require("../chunk-YHCFGHQV.js");
var import_chunk_CTBG5EFY = require("../chunk-CTBG5EFY.js");
